[FG:InPlacePodVerticalScaling] Improve allocated resources checkpointing

felipeagger opened the pull request:
When the pod allocation is updated, the Kubelet always calls set allocation at the pod level, but the status manager sets it for _each container,_ calling through to the checkpoint state. This causes the checkpoint file to be written in quick succession for each container in the pod. It would be better to just update the checkpoint once per-pod.

Changed calls to set allocation from container level to _pod level_ on status manager.

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

Fixes #128188

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources
```

ffromani commented:
/triage accepted
/priority important-longterm
/ok-to-test

priority set to match the linked issue, subject to possible refinement in the future

felipeagger commented:
/test pull-kubernetes-integration

hshiina commented:
Does this PR address only the first item in #128188? If so, it would be better not to use `Fixes` in the description because the issue will be closed when this PR is merged.

hshiina commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -296,13 +296,14 @@ func (m *manager) GetPodResizeStatus(podUID types.UID) v1.PodResizeStatus {
 func (m *manager) SetPodAllocation(pod *v1.Pod) error {
 	m.podStatusesLock.RLock()
```
Though this is not directly related to this issue, I wonder why a read lock is taken for writing a checkpoint.

felipeagger replied:
i believe, this is necessary because other methods write to this status, and this method reads from the status, so this would be necessary to avoid data race.

felipeagger commented:
> Does this PR address only the first item in #128188? If so, it would be better not to use `Fixes` in the description because the issue will be closed when this PR is merged.

The idea its fix entire issue #128188... I appreciate feedback/ideas on how to resolve item 2.

hshiina commented:
> > Does this PR address only the first item in #128188? If so, it would be better not to use `Fixes` in the description because the issue will be closed when this PR is merged.
> 
> The idea its fix entire issue #128188... I appreciate feedback/ideas on how to resolve item 2.

I think we can change here not to return an error but to log the error:
https://github.com/kubernetes/kubernetes/blob/4114a9b4e45a4df96f0383d87b2649640a6ffbf1/pkg/kubelet/kubelet.go#L2923-L2926
like here:
https://github.com/kubernetes/kubernetes/blob/4114a9b4e45a4df96f0383d87b2649640a6ffbf1/pkg/kubelet/kubelet.go#L2659-L2663

felipeagger commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -296,13 +296,14 @@ func (m *manager) GetPodResizeStatus(podUID types.UID) v1.PodResizeStatus {
 	m.podStatusesLock.RLock()
 	defer m.podStatusesLock.RUnlock()
+
+	podAlloc := state.PodResourceAllocation{}
```
Great point, I had forgotten. Solved!

felipeagger commented:
> > > Does this PR address only the first item in #128188? If so, it would be better not to use `Fixes` in the description because the issue will be closed when this PR is merged.
> > 
> > 
> > The idea its fix entire issue #128188... I appreciate feedback/ideas on how to resolve item 2.
> 
> I think we can change here not to return an error but to log the error:
> 
> https://github.com/kubernetes/kubernetes/blob/4114a9b4e45a4df96f0383d87b2649640a6ffbf1/pkg/kubelet/kubelet.go#L2923-L2926
> 
> 
> like here:
> https://github.com/kubernetes/kubernetes/blob/4114a9b4e45a4df96f0383d87b2649640a6ffbf1/pkg/kubelet/kubelet.go#L2659-L2663

perfect, done!

tallclair commented:
/assign

tallclair commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -296,13 +296,14 @@ func (m *manager) GetPodResizeStatus(podUID types.UID) v1.PodResizeStatus {
 	}
-	return nil
+
+	return m.state.SetPodResourceAllocation(podAlloc)
```
Agreed.

This method name is confusing. It's only used by restore state in the checkpoint state. You could have `restoreState` return the pod allocation, and then modify the constructor `NewStateMemory` to accept the initial allocation, and then repurpose `SetPodResourceAllocation(pod)` to set the single-pod allocation. Does that make sense? If you go this route, I'd prefer the refactoring to be a separate PR.

felipeagger replied:
Nice, Make sense! I'll going to follow this route! 

How far can we go in this PR? since the other part would be in another PR.

felipeagger committed:
kubelet: improve allocated resources checkpointing

changed calls to set allocation from container level to pod level on status manager.

felipeagger commented on pkg/kubelet/status/state/state_checkpoint.go:
```diff
@@ -76,10 +76,14 @@ func (sc *stateCheckpoint) restoreState() error {
-	err = sc.cache.SetPodResourceAllocation(praInfo.AllocationEntries)
-	if err != nil {
-		return fmt.Errorf("failed to set pod resource allocation: %w", err)
+
```
This was only necessary to avoid breaking it. In this [PR](https://github.com/kubernetes/kubernetes/pull/129504) this call **SetPodResourceAllocation** is removed from here.

felipeagger commented:
/test pull-kubernetes-node-kubelet-podresize

tallclair commented on pkg/kubelet/kubelet.go:
```diff
@@ -2922,7 +2922,7 @@ func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod, podStatus *kubecontaine
 		// Update pod resource allocation checkpoint
 		if err := kl.statusManager.SetPodAllocation(pod); err != nil {
-			return nil, err
+			klog.ErrorS(err, "SetPodAllocation failed", "pod", klog.KObj(pod))
```
I changed my mind about this change. I think we still want to fail the resize if the allocation can't be checkpointed, otherwise the node could become unstable on restart. Let's revert this line to what we had before.

tallclair commented on pkg/kubelet/status/state/state_mem.go:
```diff
@@ -74,12 +74,19 @@ func (s *stateMemory) SetContainerResourceAllocation(podUID string, containerNam
 
-	s.podAllocation = a.Clone()
-	klog.V(3).InfoS("Updated pod resource allocation", "allocation", a)
+	if _, ok := s.podAllocation[podUID]; !ok {
+		s.podAllocation[podUID] = make(map[string]v1.ResourceRequirements)
+	}
+
+	for containerName, containerAlloc := range alloc {
+		s.podAllocation[podUID][containerName] = containerAlloc
+	}
```
can't we just do
```go
s.podAllocation[podUID] = alloc
```
here? Any reason we need to deep-copy the alloc map?

tallclair replied:
I think this comment still stands? 

```suggestion
	s.podAllocation[podUID] = alloc
```

felipeagger replied:
perfect, adjusted!

felipeagger commented:
/test pull-kubernetes-e2e-gce

felipeagger commented:
/test pull-kubernetes-node-kubelet-podresize

felipeagger commented:
/test pull-kubernetes-e2e-gce

felipeagger commented:
/test pull-kubernetes-node-kubelet-podresize

felipeagger commented:
/test pull-kubernetes-e2e-gce

felipeagger commented:
/test pull-kubernetes-node-kubelet-podresize

tallclair commented:
/lgtm
/approve
